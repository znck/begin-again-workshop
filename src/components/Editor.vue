<template>
  <div class="editor">
    <textarea class="input" placeholder="Start writing..." v-model="content" />
  </div>
</template>


<script>
export default {
  name: 'Editor',
  props: {
    value: {
      type: String,
      required: true
    }
  },
  computed: {
    content: {
      get () {
        return this.value
      },
      set (value) {
        this.$emit('input', value)
      }
    }
  }
}
</script>

<style scoped>
.editor, .input {
  width: 100%;
}

.input {
  height: 100%;
  min-height: 300px;
  border: 0;
  font-family: monospace;
  font-size: 1rem;
  line-height: 1.5;
  resize: none;
}

.input:focus {
  outline: none;
}
</style>

