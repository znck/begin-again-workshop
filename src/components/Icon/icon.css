.icon {
  cursor: pointer;
  padding: 0 8px;
  zoom: 1.5;
}
