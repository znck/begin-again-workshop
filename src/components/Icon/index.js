import Back from './Back.vue'
import Edit from './Edit.vue'
import Preview from './Preview.vue'
import Save from './Save.vue'
import Trash from './Trash.vue'

export {
  Back,
  Edit,
  Preview,
  Save,
  Trash
}
