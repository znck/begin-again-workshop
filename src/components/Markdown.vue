<template>
  <div class="output" v-html="html" />
</template>


<script>
import marked from 'marked'
export default {
  name: 'Markdown',
  props: {
    content: {
      type: String,
      required: true
    }
  },
  computed: {
    html() {
      const content = this.content.replace(/:([a-z+-][^:\s]+):/gi, (_, it) => 
      ` ![${it} emoji](https://raw.githubusercontent.com/WebpageFX/emoji-cheat-sheet.com/master/public/graphics/emojis/${it}.png)`)

      return marked(content, { sanitize: true, gfm: true, emoji: true })
    }
  }
}
</script>

<style scoped>
.output {
  padding: 15px;
}
</style>

<style>
img[alt$="emoji"] {
  height: 1em;
  width: auto;
  display: inline-block;
}
</style>

