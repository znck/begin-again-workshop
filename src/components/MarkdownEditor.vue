<template>
  <div class="markdown-editor">
    <header class="header">
      <slot></slot>
    </header>
    <main class="main">
      <Editor class="editor" v-model="content" />
      <div v-if="preview" class="sep" />
      <transition name="slide-in">
        <Markdown v-if="preview" class="preview" :content="content" />
      </transition>
    </main>
    <footer class="footer">
      <div class="stats">
        Characters: {{stats.chars}}
        Words: {{stats.words}}
      </div>

      <div class="meta">
        <slot name="meta"></slot>
      </div>
    </footer>
  </div>
</template>


<script>
import Editor from './Editor.vue'
import Markdown from './Markdown.vue'

export default {
  name: 'MarkdownEditor',
  props: {
    preview: {
      type: Boolean,
      default: true
    },
    value: {
      type: String,
      required: true
    }
  },
  components: { Editor, Markdown },
  computed: {
    content: {
      get() { return this.value || '' },
      set(value) { return this.$emit('input', value) }
    },
    stats() {
      return {
        chars: this.content.length,
        words: this.content.split(/[\r\n\s#.]+/).length
      }
    }
  }
}
</script>

<style scoped>
.markdown-editor {
  display: flex;
  flex-direction: column;
  min-height: 100%;
}

.header, .footer {
  background: #efefef;
  padding: 8px 16px;
  display: flex;
  flex-direction: row;
  align-items: center;
}

.main {
  display: flex;
  flex-direction: row;
}

.main, .editor, .preview, .meta {
  flex: 1;
}

.editor {
  padding: 16px;
}

.sep {
  border: 1px solid #efefef;
}

/* Slide in animation */
.slide-in-enter-active, .slide-in-leave-active {
  transition: transform .5s;
  transform: translateX(0)
}
.slide-in-enter, .slide-in-leave-to {
  transform: translateX(100%)
}
</style>
