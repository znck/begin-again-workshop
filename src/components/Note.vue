<template>
  <section class="note">
    <Markdown class="preview" :content="body" />
    <h2 class="title">{{ title }}</h2>
  </section>
</template>

<script>
import Markdown from './Markdown.vue'

export default {
  name: 'NotePreview',
  props: {
    title: {
      type: String,
      required: true
    },
    body: {
      type: String,
      required: true
    }
  },
  components: { Markdown },
}
</script>

<style scoped>
.note {
  height: 320px;
  overflow: hidden;
  position: relative;
  margin: 16px;
  box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14), 0 1px 5px 0 rgba(0, 0, 0, 0.12), 0 3px 1px -2px rgba(0, 0, 0, 0.2);
}

.title {
  position: absolute;
  background: white;
  padding: 16px;
  white-space: nowrap;
  overflow: hidden;
  text-overflow: ellipsis;
  left: 0;
  right: 0;
  bottom: 0;
  margin: 0;
  text-align: center;
}

.preview {
  font-size: 8px;
  height: 100%;
  background: #f3f3f4;
}
</style>

