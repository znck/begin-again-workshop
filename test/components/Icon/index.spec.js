import { shallow } from '@vue/test-utils'
import { createRenderer } from 'vue-server-renderer'
import * as Icons from '@/components/Icon'

const names = Object.keys(Icons)
const renderer = createRenderer({})

names.forEach(
  name => describe(`<${name}Icon>`, () => {
    test(`snapshot`, (done) => {
      const wrapper = shallow(Icons[name])
      
      renderer.renderToString(wrapper.vm, (err, str) => {
        expect(str).toMatchSnapshot()
        done()
      })
    })
  })
)