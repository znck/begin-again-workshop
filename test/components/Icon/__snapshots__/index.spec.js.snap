// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BackIcon> snapshot 1`] = `
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="5" height="14" viewBox="0 0 5 14" class="icon">
    <path d="M4.898 4.25q0 0.102-0.078 0.18l-3.070 3.070 3.070 3.070q0.078 0.078 0.078 0.18t-0.078 0.18l-0.391 0.391q-0.078 0.078-0.18 0.078t-0.18-0.078l-3.641-3.641q-0.078-0.078-0.078-0.18t0.078-0.18l3.641-3.641q0.078-0.078 0.18-0.078t0.18 0.078l0.391 0.391q0.078 0.078 0.078 0.18z"></path>
</svg>
`;

exports[`<EditIcon> snapshot 1`] = `
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="12" height="14" viewBox="0 0 12 14" class="icon">
    <path d="M3.156 8.656l1.187 1.187-0.406 0.406h-0.438v-0.75h-0.75v-0.438zM6.391 5.609q0.109 0.102-0.023 0.234l-2.273 2.273q-0.133 0.133-0.234 0.023-0.109-0.102 0.023-0.234l2.273-2.273q0.133-0.133 0.234-0.023zM4.25 11l4.25-4.25-2.25-2.25-4.25 4.25v2.25h2.25zM9 6.25l0.719-0.719q0.219-0.219 0.219-0.531t-0.219-0.531l-1.187-1.187q-0.219-0.219-0.531-0.219t-0.531 0.219l-0.719 0.719zM12 3.25v7.5q0 0.93-0.66 1.59t-1.59 0.66h-7.5q-0.93 0-1.59-0.66t-0.66-1.59v-7.5q0-0.93 0.66-1.59t1.59-0.66h7.5q0.93 0 1.59 0.66t0.66 1.59z"></path>
</svg>
`;

exports[`<PreviewIcon> snapshot 1`] = `
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="13" height="14" viewBox="0 0 13 14" class="icon">
    <path d="M1.25 12h4.75v-9h-5v8.75q0 0.102 0.074 0.176t0.176 0.074zM12 11.75v-8.75h-5v9h4.75q0.102 0 0.176-0.074t0.074-0.176zM13 2.25v9.5q0 0.516-0.367 0.883t-0.883 0.367h-10.5q-0.516 0-0.883-0.367t-0.367-0.883v-9.5q0-0.516 0.367-0.883t0.883-0.367h10.5q0.516 0 0.883 0.367t0.367 0.883z"></path>
</svg>
`;

exports[`<SaveIcon> snapshot 1`] = `
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="12" height="14" viewBox="0 0 12 14" class="icon">
    <path d="M3 12h6v-3h-6v3zM10 12h1v-7q0-0.109-0.078-0.301t-0.156-0.27l-2.195-2.195q-0.078-0.078-0.266-0.156t-0.305-0.078v3.25q0 0.312-0.219 0.531t-0.531 0.219h-4.5q-0.312 0-0.531-0.219t-0.219-0.531v-3.25h-1v10h1v-3.25q0-0.312 0.219-0.531t0.531-0.219h6.5q0.312 0 0.531 0.219t0.219 0.531v3.25zM7 4.75v-2.5q0-0.102-0.074-0.176t-0.176-0.074h-1.5q-0.102 0-0.176 0.074t-0.074 0.176v2.5q0 0.102 0.074 0.176t0.176 0.074h1.5q0.102 0 0.176-0.074t0.074-0.176zM12 5v7.25q0 0.312-0.219 0.531t-0.531 0.219h-10.5q-0.312 0-0.531-0.219t-0.219-0.531v-10.5q0-0.312 0.219-0.531t0.531-0.219h7.25q0.312 0 0.688 0.156t0.594 0.375l2.188 2.187q0.219 0.219 0.375 0.594t0.156 0.688z"></path>
</svg>
`;

exports[`<TrashIcon> snapshot 1`] = `
<svg version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" width="11" height="14" viewBox="0 0 11 14" class="icon">
    <path d="M4 5.75v4.5q0 0.109-0.070 0.18t-0.18 0.070h-0.5q-0.109 0-0.18-0.070t-0.070-0.18v-4.5q0-0.109 0.070-0.18t0.18-0.070h0.5q0.109 0 0.18 0.070t0.070 0.18zM6 5.75v4.5q0 0.109-0.070 0.18t-0.18 0.070h-0.5q-0.109 0-0.18-0.070t-0.070-0.18v-4.5q0-0.109 0.070-0.18t0.18-0.070h0.5q0.109 0 0.18 0.070t0.070 0.18zM8 5.75v4.5q0 0.109-0.070 0.18t-0.18 0.070h-0.5q-0.109 0-0.18-0.070t-0.070-0.18v-4.5q0-0.109 0.070-0.18t0.18-0.070h0.5q0.109 0 0.18 0.070t0.070 0.18zM9 11.406v-7.406h-7v7.406q0 0.172 0.055 0.316t0.113 0.211 0.082 0.066h6.5q0.023 0 0.082-0.066t0.113-0.211 0.055-0.316zM3.75 3h3.5l-0.375-0.914q-0.055-0.070-0.133-0.086h-2.477q-0.078 0.016-0.133 0.086zM11 3.25v0.5q0 0.109-0.070 0.18t-0.18 0.070h-0.75v7.406q0 0.648-0.367 1.121t-0.883 0.473h-6.5q-0.516 0-0.883-0.457t-0.367-1.105v-7.438h-0.75q-0.109 0-0.18-0.070t-0.070-0.18v-0.5q0-0.109 0.070-0.18t0.18-0.070h2.414l0.547-1.305q0.117-0.289 0.422-0.492t0.617-0.203h2.5q0.312 0 0.617 0.203t0.422 0.492l0.547 1.305h2.414q0.109 0 0.18 0.070t0.070 0.18z"></path>
</svg>
`;
